### ISSUE TITLE
unix packaging tests » ubuntu-22.04-packaging :distribution:docker:buildDockerImage failing

### ISSUE THREAD

**Issue by n1v0lg** [elastic/elasticsearch#96207]
### CI Link

https://gradle-enterprise.elastic.co/s/djj3txgj25vic

### Repro line

N/A

### Does it reproduce?

Didn't try

### Applicable branches

main, 8.7

### Failure excerpt

I've having a hard time making sense of this one:

```
| #5 ERROR: process "/bin/sh -c yes no | dpkg-reconfigure dash &&     for iter in 1 2 3 4 5 6 7 8 9 10; do       export DEBIAN_FRONTEND=noninteractive &&       apt-get update &&       apt-get upgrade -y &&       apt-get install -y --no-install-recommends         ca-certificates curl netcat p11-kit unzip zip  &&       apt-get clean &&       rm -rf /var/lib/apt/lists/* &&       exit_code=0 && break ||         exit_code=$? && echo \"apt-get error: retry $iter in 10s\" && sleep 10;     done;     exit $exit_code" did not complete successfully: exit code: 4294967295
```

Is the first error occurrence. Further down again:

```
| Dockerfile:84
| --------------------
|   83 |     # Change default shell to bash, then install required packages with retries.
|   84 | >>> RUN yes no | dpkg-reconfigure dash && \
|   85 | >>>     for iter in 1 2 3 4 5 6 7 8 9 10; do \
|   86 | >>>       export DEBIAN_FRONTEND=noninteractive && \
|   87 | >>>       apt-get update && \
|   88 | >>>       apt-get upgrade -y && \
|   89 | >>>       apt-get install -y --no-install-recommends \
|   90 | >>>         ca-certificates curl netcat p11-kit unzip zip  && \
|   91 | >>>       apt-get clean && \
|   92 | >>>       rm -rf /var/lib/apt/lists/* && \
|   93 | >>>       exit_code=0 && break || \
|   94 | >>>         exit_code=$? && echo "apt-get error: retry $iter in 10s" && sleep 10; \
|   95 | >>>     done; \
|   96 | >>>     exit $exit_code
|   97 |     
| --------------------
| ERROR: failed to solve: process "/bin/sh -c yes no | dpkg-reconfigure dash &&     for iter in 1 2 3 4 5 6 7 8 9 10; do       export DEBIAN_FRONTEND=noninteractive &&       apt-get update &&       apt-get upgrade -y &&       apt-get install -y --no-install-recommends         ca-certificates curl netcat p11-kit unzip zip  &&       apt-get clean &&       rm -rf /var/lib/apt/lists/* &&       exit_code=0 && break ||         exit_code=$? && echo \"apt-get error: retry $iter in 10s\" && sleep 10;     done;     exit $exit_code" did not complete successfully: exit code: 4294967295
```

Apologies if this is a false alarm! I did not get to dig too deeply here but it does not seem like a flaky network issue.

**Comment by mark-vieira** [comment]
It's not clear what the error actually is here. I'm not sure if it's just because nothing is being output, or our retry mechanism is swallowing the error. I seems like whatever this is is transient as it's not consistently happening.

**Comment by mark-vieira** [comment]
This is definitely a continuing problem and is happening several times a day. It's unclear from the console output what exactly his happening here. The common bit seems that this is happening where we call `apt-get` and that these calls are wrapped in some retry logic. What's not obvious is if there is something funny with the retry logic or the package manager. I don't see any error codes, and we don't seem to be actually performing a retry.

I'm apt (pun intended) to try tripping out the retry stuff to see if that either a) sorts out the problem or b) gives us some better output.

**Comment by mark-vieira** [comment]
I've created https://github.com/elastic/elasticsearch/pull/96256 to disable the retry logic for now in `main` to see if that provides some better log output in these failures.

**Comment by mark-vieira** [comment]
I haven't seen any instances of this failure in `main` since merging #96256. Related? Maybe. I'll let it soak for another day or two.

**Comment by mark-vieira** [comment]
And here's one that did fail with the retry omitted. Still no helpful message. Just that the command failed with a useless error code.

https://gradle-enterprise.elastic.co/s/ztn3bcuw7zwhy/console-log?task=:distribution:docker:buildAarch64DockerImage

I'm struggling to see how to press forward here. It's almost as though the build processes are being oom killed or something. I'll try and dig into the agent logs to see if anything stands out.

**Comment by pugnascotia** [comment]
That error code seems to come up a lot with Docker on Windows, and there's also this issue: https://github.com/moby/moby/issues/40844

I do wonder if there's something funky going on with Docker at this point.

**Comment by pugnascotia** [comment]
There's also https://github.com/moby/moby/issues/36682

**Comment by mark-vieira** [comment]
Thanks @raoulbhatia. Do you know when you started observing these errors? We've noticed them start around a week ago.

I've done some digging through agent syslogs and nothing stands outs. I've compared syslog output of successful builds to failed ones and I see all the same output.

I've ruled out a few things:

1. It doesn't seem specific to things like `apt-get`. Here's an [example of a command](https://gradle-enterprise.elastic.co/s/7v5ozq52senda/console-log?page=2&task=:distribution:docker:buildDockerImage) that involves zero network IO failing. My guess is the reason we see more errors in stuff like `apt-get` commands is that they are simply longer running so more likely to overlap with whatever the root issue is.
2. It's not related to parallel `docker build` tasks. In [this same build](https://gradle-enterprise.elastic.co/s/7v5ozq52senda/console-log?page=2&task=:distribution:docker:buildDockerImage) there was only one docker build running and we still exploded.
3. It's not related to buildx or our usage of qemu for buliding ARM images as we see errors building both x86 and aarch64 images.
4. It doesn't seem to be specific to any particular platform. I've seen these errors across the entire range of Linux distributions we support.

So yeah, not a lot to go on here quite yet unfortunately. I'm going to try add some diagnostics to capture docker daemon logs to see if perhaps that might give us some more insight.

**Comment by raoulbhatia** [comment]
@mark-vieira quick checks point to the first failure around May 17, so 1 week ago like you mentioned as well.

Like you, I am currently not able to make out a distinct pattern.  I'll continue to investigate tomorrow and will report any finding.

FYI we are running on Linux — I am not aware of any Windows-related component in these deployments.

**Comment by mark-vieira** [comment]
> FYI we are running on Linux — I am not aware of any Windows-related component in these deployments.

Same. We aren't doing anything Docker-related on Windows.

**Comment by raoulbhatia** [comment]
@mark-vieira a quick update from my side - pinning Docker to 23* makes the pipelines happy again.
I am using the following code on Debian-based distros, where `1001` enforces a downgrade, if required [1].

```
cat > /etc/apt/preferences.d/docker-pin.pref <<EOF
Package: docker-ce*
Pin: version 5:23*
Pin-Priority: 1001
EOF
```

[1] https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#listofnotablepinpinningtechnique

**Comment by raoulbhatia** [comment]
@mark-vieira will you report this issue to Docker / Moby?

**Comment by mark-vieira** [comment]
> @mark-vieira will you report this issue to Docker / Moby?

That's a secondary concern at this point. I'd have to figure out _what_ to report. I'll have to see if I can reliably reproduce this and get additional logging. I suspect the absolute lack of any useful diagnostics would be as useless to them as it has been to me.

**Comment by mark-vieira** [comment]
I have an open pull request to pin the docker version on our CI agents to 23.0 for now. Let's see if that resolves it.

**Comment by n1v0lg** [comment]
@mark-vieira I think we have a new failure here: https://gradle-enterprise.elastic.co/s/3yalw7ogkvqmc

```
| #12 ERROR: process "/bin/sh -c tar -zxf /tmp/elasticsearch.tar.gz --strip-components=1" did not complete successfully: exit code: 4294967295
| 
| #6 [stage-1  2/10] RUN yes no | dpkg-reconfigure dash &&     for iter in 1 2 3 4 5 6 7 8 9 10; do       export DEBIAN_FRONTEND=noninteractive &&       apt-get update &&       apt-get upgrade -y &&       apt-get install -y --no-install-recommends         ca-certificates curl netcat p11-kit unzip zip &&       apt-get clean &&       rm -rf /var/lib/apt/lists/* &&       exit_code=0 && break ||         exit_code=$? && echo "apt-get error: retry $iter in 10s" && sleep 10;     done;     exit $exit_code
| #6 28.23 Setting up libasn1-8-heimdal:amd64 (7.7.0+dfsg-1ubuntu1.4) ...
| #6 28.25 Setting up netcat-openbsd (1.206-1ubuntu1) ...
| #6 CANCELED
| ------
|  > [builder  7/10] RUN tar -zxf /tmp/elasticsearch.tar.gz --strip-components=1:
| ------
| Dockerfile:47
| --------------------
|   45 |     COPY elasticsearch-7.17.11-SNAPSHOT-linux-x86_64.tar.gz /tmp/elasticsearch.tar.gz
|   46 |     
|   47 | >>> RUN tar -zxf /tmp/elasticsearch.tar.gz --strip-components=1
|   48 |     
|   49 |     # The distribution includes a `config` directory, no need to create it
| --------------------
| ERROR: failed to solve: process "/bin/sh -c tar -zxf /tmp/elasticsearch.tar.gz --strip-components=1" did not complete successfully: exit code: 4294967295
```

It's on 7.17 -- the [fix](https://github.com/elastic/infra/pull/39136/files) you merged should still apply though right?

**Comment by bpintea** [comment]
Should it help, a fresh one (tho on main): https://gradle-enterprise.elastic.co/s/5uwgph25tboxe

**Comment by raoulbhatia** [comment]
Hi!

Did you manage to find a solution to this issue?
Are you still running with Docker `23.*` or were you able to use `24.*`?

Thanks,
Raoul

**Comment by Denton-L** [comment]
FWIW, it looks like [24.0.2](https://docs.docker.com/engine/release-notes/24.0/#2402) fixes some `exit code: 4294967295` bug. Not sure if it's the same one but it might be worth upgrading to see if it fixes anything.

**Comment by mark-vieira** [comment]
We did eventually sort this out by skipping that bad Docker version.
